src/iop/useless.c

```diff
@@ -41,17 +60,24 @@ typedef struct dt_iop_useless_params_t
   // stored in self->params and self->default_params
   // also, since this is stored in db, you should keep changes to this struct
   // to a minimum. if you have to change this struct, it will break
-  // users data bases, and you should increment the version
-  // of DT_MODULE(VERSION) above!
-  int checker_scale;
-  float factor;
+  // users data bases, and you have to increment the version
+  // of DT_MODULE(VERSION) above and provide a legacy_params upgrade path!
+  // Tags in the comments get picked up by the introspection framework and are
+  // used to set range and default for widgets in gui_init, labels (for widgets and history)
+  // and value checks before commit_params.
+  // If no explicit init() is specified, the default implementation uses $DEFAULT tags
+  // to initialise self->default_params, which is then used in gui_init to set widget defaults.
+  int checker_scale; // $MIN: 0 $MAX: 10 $DEFAULT: 1 $DESCRIPTION: "size"
+  float factor;      // $MIN: -5.0 $MAX: 5.0 $DEFAULT: 0
+  gboolean check;    // $DESCRIPTION: "checkbox option"
+  dt_iop_useless_type_t method; // $DEFAULT: DT_USELESS_SECOND $DESCRIPTION: "parameter choices"
 } dt_iop_useless_params_t;
 
 typedef struct dt_iop_useless_gui_data_t
 {
   // whatever you need to make your gui happy.
   // stored in self->gui_data
-  GtkWidget *scale, *factor; // this is needed by gui_update
+  GtkWidget *scale, *factor, check, method, extra; // this is needed by gui_update
```

Comment:
Similarly be clear that these are examples.

Reply:
see above. we are not aiming for the very lowest common denominator here; unless I'm wrong, anybody that has any chance of understanding any of this will get that these are examples? from a writer's perspective you are absolutely right of course (and if you give me concrete sentence to insert I'll do it!) but my brain just can't think of anything to add without feeling silly ;-)

Reply:
As above, possibly well chosen variable names would make it clear.